core/src/main/scala/org/apache/spark/executor/ExecutorMetricsPoller.scala

```diff
@@ -107,8 +108,8 @@ private[spark] class ExecutorMetricsPoller(
    * Called by TaskRunner#run.
    */
   def onTaskStart(taskId: Long, stageId: Int, stageAttemptId: Int): Unit = {
-    // Put an entry in taskMetricPeaks for the task.
-    taskMetricPeaks.put(taskId, new AtomicLongArray(ExecutorMetricType.numMetrics))
+    // Put an entry in taskMetricPeaks for the task with latestMetrics.
```

Comment:
I guess your issue is the metrics didn't get updated because `poll()` was not triggered? In that case why not update metrics snapshot here?

Reply:
> I guess your issue is the metrics didn't get updated because `poll()` was not triggered? In that case why not update metrics snapshot here or `onTaskCompletion`?

The issue is during task start and start end, metric poller don't call poll() if task during less 
 then poll interval. then in task completion events, task's executor metrics will be sent with initial 0 value.

Reply:
> In that case why not update metrics snapshot here or `onTaskCompletion`?

snapshot will sent to driver in heartbeat and create a ExecutorMetricsUpdate Event, but heartbeat interval is always bigger then task duration, the final value is still not updated.

Reply:
you should get the latest metrics here right? otherwise the metrics might be stale.